pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,424 @@
+	}
+
+	server, _, err := client.Server.Create(c.ProjectID, &serverCreateRequest)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to create server = %v", err),
+		}
+	}
+
+	return &cherryServer{server: server}, nil
+}
+
+type cherryServer struct {
+	server cherrygo.Server
+}
+
+func (s *cherryServer) Name() string {
+	return s.server.Hostname
+}
+
+func (s *cherryServer) ID() string {
+	return strconv.Itoa(s.server.ID)
+}
+
+func (s *cherryServer) Addresses() []string {
+	var addresses []string
+	for _, ip := range s.server.IPAddresses {
+		addresses = append(addresses, ip.Address)
+	}
+	return addresses
+}
+
+func (s *cherryServer) Status() instance.Status {
+	if s.server.State != "active" {
+		return instance.StatusCreating
+	}
+	return instance.StatusRunning
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	servers, _, err := client.Servers.List(c.ProjectID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to list servers, due to %v", err),
+		}
+	}
+
+	for _, server := range servers {
+		if server.Tags[machineUIDLabelKey] == string(machine.UID) {
+			srv, _, err := client.Server.List(strconv.Itoa(server.ID))
+			if err != nil {
+				return nil, cloudprovidererrors.TerminalError{
+					Reason:  common.InvalidConfigurationMachineError,
+					Message: fmt.Sprintf("Failed to fetch server, due to %v", err),
+				}
+			}
+			return &cherryServer{server: srv}, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, new types.UID) error {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+	client := getClient(c.Token)
+
+	server, err := getServerByName(client, machine.Name, c.ProjectID)
+	if err != nil {
+		return fmt.Errorf("failed to get server: %v", err)
+	}
+	if server == nil {
+		glog.Infof("No instance exists for machine %s", machine.Name)
+		return nil
+	}
+
+	glog.Infof("Setting UID label for machine %s", machine.Name)
+	updateServerRequest := cherrygo.UpdateServer{
+		Tags: map[string]string{machineUIDLabelKey: string(new)},
+	}
+
+	_, response, err := client.Server.Update(strconv.Itoa(server.ID), &updateServerRequest)
+	if err != nil {
+		return fmt.Errorf("failed to update UID label: %v", err)
```

Comment:
We should eventually return the TerminalError here.

Reply:
Fixed